text/0000-disconnect-flow.md

```diff
@@ -0,0 +1,134 @@
+- Feature Name: `disconnect_flow`
+- Start Date: 2021/06/02
+- RFC PR: 
+- Issue: [[Unity-Technologies/com.unity.multiplayer.mlapi#772](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/issues/772)
+
+# Summary
+
+This feature standardizes the flow for communicating disconnects, both between server and client, and between SDK and user.
+
+# Motivation
+
+The motivation is to better enable users to react to network events, including servers reacting to disconnecting clients, clients reacting to lost connections to the server, and clients reacting to other clients joining and leaving the game.
+
+# Guide-level explanation
+
+When network topology changes (via new or lost connections, whether intentionally or unintentionally), it's often desirable for the processes that are (or were) involved in the network to be informed of it so that they can react. In order to reactto these events, MLAPI provides the following callbacks that can be registered to NetworkManager:
+
+- `OnClientConnectedCallback` - Called on the **server** when a new client connection is **formed**. Expects a delegate with the following signature:
+  
+  ```C#
+  void OnClientConnected(int clientId){}
+  ```
+
+- `OnClientDisconnectCallback` - Called on the **server** when a client connection is **lost**. Expects a delegate with the following signature:
+  
+  ```C#
+  void OnClientDisconnected(int clientId, NetworkManager.DisconnectReason reason)
+  ```
+
+- `OnPeerConnectedCallback` - Called on the **client** when another client has **connected** to the same server. Expects a delegate with the following signature:
+  
+  ```C#
+  void OnPeerConnected(int clientId){}
+  ```
+
+- `OnPeerDisconnectedCallback` - Called on the **client** when another client has **disconnected** from the same server. Expects a delegate with the following signature:
+  
+  ```C#
+  void OnPeerDisconnected(int clientId, NetworkManager.DisconnectReason reason)
+  ```
+
+- `OnServerConnectionEstablishedCallback` - Called on the **client** when a server connection is **established**. Expects a delegate with the following signature, where `peers` is the list of clients who are already in the game:
+  
+  ```C#
+  void OnServerConnectionEstablished(int[] peers){}
+  ```
+
+- `OnServerConnectionLostCallback` - Called on the **client** when it is **no longer connected** to the server (i.e., no longer a part of the game at all). Expects a delegate with the following signature:
+  
+  ```C#
+  void OnServerConnectionLost(NetworkManager.DisconnectReason reason)
+  ```
+
+The `reason` parameter is an enum currently defined as follows:
+
+```C#
+public enum DisconnectReason
```

Comment:
I'm wondering if we could get inspired by HTTP error codes here.
for example, they have 4xx error codes which are client errors (asking for the wrong things or the wrong way) --> 400 is malformed request, 401 is unauthorized (where auth failed), 403 is when auth worked, but the user is forbidden from accessing, 404 for ressource not found, etc.
Same for 5xx error codes. We could have 500 for a reason where the server had an exception and the client couldn't connect (if there's any exceptions thrown server side during connection approval for example). Or 503 when the server was supposed to be there, but unavailable, signalling to the client to try again later (maybe using exponential backoff for example)
I don't think they are all applicable, but they might be a great inspiration for an exhaustive list of reasons?

Reply:
I think that's valuable. I think I would prefer, personally, to still see them represented as an enum, but having the enum values logically grouped would help with debugging, and we could follow the same prefix as HTTP just to make it more familiar. 400-499 is something wrong with the request (not authorized, malformed connect request, version mismatch, etc), 500-599 is something wrong with the request handler (unhandled exception in authorization callback, time out, etc)

Reply:
oh you're right, I wasn't suggesting we use ints for these error codes. I don't really mind ints vs enums.
It's mostly about the list of errors we can get inspiration from. HTTP has been established for a while, it has a pretty good list of use cases we might want to express too.
Good idea on the user extensible values.